README.rst
ABOUT THE MODULE
================
AwstatsReader is a pythonic interface to AWStats data cache files.  Using it,
you can access year and month via dict-like subscripts, and and individual
data points via both dict-like subscripts and attribute-like accessors.

As of version 0.5, it includes a script for merging AWStats Cache files.

Download: http://azariah.com/open_source.html

ABOUT THE AUTHOR
================
Joshua Kugler (joshua@azariah.com) is a programmer and system administator
with over 10 years of industory experience.
Resume at: http://jjncj.com/papers/KuglerResume.pdf

DISCLAIMER
==========
This is an beta-ish release.  There are 43 tests which cover most, if not
all of the functionality, but not much documentation.  The interface should be
considered stable, but not in concrete.  The usage of this project in a "real
world" situation (awstats_cache_merge.py) led to many improvements to the API.

I wrote this via examples from an AWStats cache file, so I'm sure there are
sections for which I do not have definitions.  If you would send me those
sections, I'll be sure to add them.

Right now, this will parse and display cache files from AWStats 6.5. I've not
tested other versions yet, as 6.5 is the only version I've had access to so far.

REPOSITORY
==========
http://github.com/jkugler/awstats_reader

INSTALLATION
============
See INSTALL

LICENSE
=======
See COPYING

EXAMPLE
=======
Usage is pretty self explanatory::

    import awstats_reader

    obj  = awstats_reader.AwstatsReader('/path/to/awstats_logs', 'example.com')

    print obj[2007]
    print obj[2008][6]
    m = obj[2009][7]
    print m['general']
    # Access like a dictionary...
    print m['general']['LastLine']
    #...or like an object attribute
    print m['general'].LastLine
    print m.general.LastLine

FEEDBACK
========
Please send questions/comments/suggestions to awstatsreader@azariah.com
For now, you can find the latest version here: http://azariah.com/open_source.html


docs/CHANGES
Change Log

+ New Feature
- Bug fix
| Neutral change
* Incompatible change

2009-12-19
  + More doc changes

2009-12-15
  + Some documenation improvements

2009-12-13
  + More tests. 43 total
  | Took dead code out of AwstatsDate
  * Changed AwstatsMonth.__str__ from YYYY/MM to YYYY-MM
  | Updated documentation

2009-12-06
  + And more tests. 32 total
  | Corrected an IOError to an OSError on file or directory not found
  | Deleted some old comments
  + Added some test cache files

2009-11-29
  + More tests. Up to 13! :)

2009-11-29
  + Test framework in place, along with a few tests

2009-11-22
  * Renamed package to awstats_reader to better comply with PEP8
  | awstats_cache_merge.py using new package name
  | Refactored the merge code back into the AwstatsSection class
  | Caught up the change log

2009-11-04
  + Tagged the repo with EE Internet's permission to release awstats_cache_merge.py as open source

2009-11-01
  + Rows are order stabilized now when merging
  + Came up with a way to handle year '0'

2009-10-31
  + Dealing with year '0'
  + Month object now has a "key()s" method
  + Improved sorting: can sort in reverse, and with integer keys


2009-10-21
  + Added awstats_cache_merge.py
  + Modifications to AwstatsReader to make merging work more smoothly

2009-08-02
  + Initial Releaese


run_tests.py
#!/usr/bin/env python

import os
import sys
import unittest

# Set up the test environment
opd = os.path.dirname
sys.path.insert(0, opd(os.path.abspath(__file__)))

from awstats_reader import tests as awsr_tests

if __name__ == '__main__':
    suite = unittest.TestLoader().loadTestsFromModule(awsr_tests)
    unittest.TextTestRunner(verbosity=2).run(suite)


tests/qad.py
#!/usr/bin/python

# My quick and dirty test script - mostly to verify operation and serve
# as examples

import sys
import AwstatsReader

obj  = AwstatsReader.AwstatsReader('/home/jkugler/tmp/awstats_logs', sys.argv[1])

print obj
print obj[2007]
print obj[2008][6]
m = obj[2009][7]
print m['general']
print m['general']['LastLine']
print m['general'].LastLine
print m.general.LastLine
print m['general']['TotalVisits']
print m['visitor']['132.115.in-addr.arpa']

for x,y in m.pagerefs.items():
    print x, m.pagerefs[x].pages, m.pagerefs[x].hits



awstats_reader/tests.py
#!/usr/bin/env python

import datetime
import os
import types
import unittest

import awstats_reader

opd = os.path.dirname

test_file_dir = os.path.join(opd(opd(os.path.abspath(__file__))), 'test_files')

class TestAwstatsHelpers(unittest.TestCase):
    """Tests various helper functions in the awstats_reader package"""

    def test_14_digit_datetime(self):
        """Ensure date/time strings are being evaluated correctly"""
        obj = awstats_reader.awstats_datetime('20091130165230')
        dt = datetime.datetime(2009, 11, 30, 16, 52, 30)
        self.assertEqual(obj, dt)

    def test_8_digit_date(self):
        """Ensure date strings are being evaluated correctly"""
        obj = awstats_reader.awstats_datetime('20091130')
        dt = datetime.date(2009, 11, 30)
        self.assertEqual(obj, dt)

    def test_14_digit_datetime_detection(self):
        """Ensure the AwstatsDateTime object is returned"""
        obj = awstats_reader.awstats_datetime('20091130165230')
        self.assertTrue(isinstance(obj, awstats_reader.AwstatsDateTime))

    def test_8_digit_date_detection(self):
        """Ensure the AwstatsDate object is returned"""
        obj = awstats_reader.awstats_datetime('20091130')
        self.assertTrue(isinstance(obj, awstats_reader.AwstatsDate))

    def test_year_year_zero_datetime_parse(self):
        """Ensure the Awstats 'year zero' is parsed correctly"""
        obj = awstats_reader.awstats_datetime('0')
        self.assertEqual(obj,datetime.datetime(1,1,1))

    def test_year_zero_datetime_output(self):
        """Ensure the Awstats 'year zero' is printed correctly"""
        obj = awstats_reader.awstats_datetime('0')
        self.assertEqual(obj.strftime('%Y%m%d%H%M%S'), '0')

    def test_datetime_invalid_string(self):
        """Ensure an invalid date/time string raises an exception"""
        self.assertRaises(RuntimeError, awstats_reader.awstats_datetime, '2009')

    def test_attr_dict(self):
        """Ensure AttrDict behaves correctly"""
        obj = awstats_reader.AttrDict([('this','that'), ('thus','those')])
        self.assertEqual(obj.thus, 'those')

class TestAwstatsReader(unittest.TestCase):
    """Tests the AwstatsReader main object"""

    def test_init(self):
        """Ensure we can initialize the class"""
        ar = awstats_reader.AwstatsReader('/tmp', 'example.com')
        self.assertTrue(isinstance(ar, awstats_reader.AwstatsReader))

    def test_invalid_year_fail(self):
        """Ensure getting an invalid year raises an exception"""
        ar = awstats_reader.AwstatsReader('/tmp', 'example.com')
        self.assertRaises(KeyError, ar.__getitem__, 9999)

    def test_valid_year(self):
        """Ensure getting a valid year returns an AwstatsYear object"""
        ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')
        obj = ar[2009]
        self.assertTrue(isinstance(obj, awstats_reader.AwstatsYear))

    def test_invalid_dir(self):
        """Ensure passing an invalid directory raises an exception"""
        self.assertRaises(OSError, awstats_reader.AwstatsReader, '/tmp/XYZ', 'example.com')

    def test_iter_operation(self):
        """Ensure AwstatsYear's __iter__ function is working"""
        ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')
        year_iter = ar.__iter__()
        self.assertTrue(isinstance(year_iter, types.GeneratorType))

    def test_found_all_years(self):
        """Ensure all years were found"""
        ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')
        self.assertEqual(ar.years, [2008,2009])

    def test_iter_all_years(self):
        """Ensure year iterator is working"""
        ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')
        self.assertEqual([ary.year for ary in ar], [2008,2009])

    def test_contains_true(self):
        """Ensure __contains__ is working positively"""
        ary = self.ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')[2009]
        self.assertTrue(11 in ary)

    def test_contains_false(self):
        """Ensure __contains__ is working negatively"""
        ary = self.ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')[2009]
        self.assertFalse(10 in ary)

    def test_str(self):
        """Ensure AwstatsReader __str__ is operating correctly"""
        ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')
        self.assertEqual(str(ar), '<AwstatsReader: 2008, 2009>')


class TestAwstatsYear(unittest.TestCase):

    def setUp(self):
        self.ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')

    def test_invalid_month(self):
        """Ensure getting an invalid month raises an exception"""
        ar_year = self.ar[2009]
        self.assertRaises(KeyError, ar_year.__getitem__, 1)

    def test_valid_month(self):
        """Ensure getting a valid month returns an AwstatsMonth object"""
        ar_month = self.ar[2009][11]
        self.assertTrue(isinstance(ar_month, awstats_reader.AwstatsMonth))

    def test_found_all_months(self):
        """Ensure all months were found"""
        ary = self.ar[2009]
        self.assertEqual(ary.months, [11,12])

    def test_iter_all_months(self):
        """Ensure month is working"""
        ary = self.ar[2009]
        self.assertEqual([arm.month for arm in ary], [11,12])

    def test_str(self):
        """Ensure AwstatsYear __str__ is operating correctly"""
        ary = self.ar[2009]
        self.assertEqual(str(ary), '<AwstatsYear 2009: 11, 12>')


class TestAwstatsMonth(unittest.TestCase):
    wanted_sections = ['general', 'time', 'visitor', 'day', 'domain',
                       'login', 'robot', 'worms', 'emailsender',
                       'emailreceiver', 'session', 'sider', 'filetypes', 'os',
                       'browser', 'screensize', 'unknownreferer',
                       'unknownrefererbrowser', 'origin', 'sereferrals',
                       'pagerefs', 'searchwords', 'keywords', 'misc', 'errors',
                       'cluster', 'sider_404', 'plugin_geoip_city_maxmind']

    def setUp(self):
        self.ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')

    def test_month_iterator(self):
        """Ensure all sections are found"""
        arm = self.ar[2009][11]
        self.assertEqual(list(arm), self.__class__.wanted_sections)

    def test_month_keys(self):
        """Ensure all sections are found"""
        arm = self.ar[2009][11]
        self.assertEqual(arm.keys(), self.__class__.wanted_sections)

    def test_get_invalid_section(self):
        """Ensure getting an invalid sections raises an exception"""
        arm = self.ar[2009][11]
        self.assertRaises(KeyError, arm.__getitem__, 'invalid_section')

    def test_get_valid_section(self):
        """Ensure getting an valid section resturns an AwstatsSection object"""
        arm = self.ar[2009][11]
        ars = arm['general']
        self.assertTrue(isinstance(ars, awstats_reader.AwstatsSection))

    def test_len(self):
        """Ensure the len() function returns the correct value"""
        arm = self.ar[2009][11]
        self.assertEqual(len(arm.keys()), len(self.__class__.wanted_sections))

    def test_str(self):
        """Ensure AwstatsMonth __str__ is operating correctly"""
        arm = self.ar[2009][11]
        self.assertEqual(str(arm), '<AwstatsMonth 2009-11>')

class TestAwstatsSection(unittest.TestCase):
    wanted_lines = ['LastLine', 'FirstTime', 'LastTime', 'LastUpdate',
                    'TotalVisits', 'TotalUnique', 'MonthHostsKnown',
                    'MonthHostsUnknown']

    def setUp(self):
        self.ar = awstats_reader.AwstatsReader(test_file_dir, 'jjncj.com')

    def test_get_valid_line(self):
        """Ensure getting a valid line returns an ordered dict"""
        ars = self.ar[2009][11]['general']
        self.assertTrue(isinstance(ars['TotalVisits'], awstats_reader.AttrDict))

    def test_get_invalid_line(self):
        """Ensure getting an invalid line raises an exception"""
        ars = self.ar[2009][11]['general']
        self.assertRaises(KeyError, ars.__getitem__, 'invalid_section')

    def test_section_iterator(self):
        """Ensure the section iterator works"""
        ars = self.ar[2009][11]['general']
        self.assertEqual(list(ars), self.__class__.wanted_lines)

    def test_section_keys(self):
        """Ensure the section keys works"""
        ars = self.ar[2009][11]['general']
        self.assertEqual(ars.keys(), self.__class__.wanted_lines)

    def test_get_return_default(self):
        """Ensure get() returns the default default value"""
        ars = self.ar[2009][11]['general']
        self.assertTrue(ars.get('invalid_row') is None)

    def test_get_return_default_provided(self):
        """Ensure get() returns the provided default value"""
        ars = self.ar[2009][11]['general']
        self.assertEqual(ars.get('invalid_row', 'ZZZZZZ'), 'ZZZZZZ')

    def test_get_return_desired_value(self):
        """Ensure get() returns the desired value"""
        ars = self.ar[2009][11]['general']
        self.assertEqual(ars.get('TotalVisits'),
                         awstats_reader.AttrDict([('value', 1475)]))

    def test_get_sort_info(self):
        """Ensure sort_info() returns the correct value"""
        ars = self.ar[2009][11]['day']
        self.assertEqual(ars.get_sort_info(), (31, 'key', True))

    def test_merge_max(self):
        """Test max merge operation"""
        ars = self.ar[2009][11]['general']
        ars2 = awstats_reader.AwstatsReader(test_file_dir,
                                            'joshuakugler.com')[2009][11]['general']
        self.assertEqual(ars.merge(ars2, 'LastLine', 'line'), 1011585)

    def test_merge_repl(self):
        """Test repl merge operation"""
        ars = self.ar[2009][11]['general']
        ars2 = awstats_reader.AwstatsReader(test_file_dir,
                                            'joshuakugler.com')[2009][11]['general']
        self.assertEqual(ars.merge(ars2, 'LastLine', 'signature'), '')

    def test_merge_min(self):
        """Test min merge operation"""
        ars = self.ar[2009][11]['general']
        ars2 = awstats_reader.AwstatsReader(test_file_dir,
                                            'joshuakugler.com')[2009][11]['general']
        self.assertEqual(ars.merge(ars2, 'FirstTime', 'first_time'),
                         awstats_reader.AwstatsDateTime(2009, 11, 1, 0, 2, 37))

    def test_merge_sum(self):
        """Test sum merge operation"""
        ars = self.ar[2009][11]['general']
        ars2 = awstats_reader.AwstatsReader(test_file_dir,
                                            'joshuakugler.com')[2009][11]['general']
        self.assertEqual(ars.merge(ars2, 'LastUpdate', 'parsed'), 1262637)

    def test_merge_latest(self):
        """Test 'latest' merge operation"""
        ars = self.ar[2009][11]['sider_404']
        ars2 = awstats_reader.AwstatsReader(test_file_dir,
                                            'joshuakugler.com')[2009][11]['sider_404']
        self.assertEqual(ars.merge(ars2, '/styles/widgets/blog-widget.css',
                                   'last_url_referer'), 'http://joshuakugler.com/')

class TestAwstatsMerge(unittest.TestCase):
    """Test functions and procedures in awstats_cache_merge"""

    def test_make_get_field(self):
        """Ensure make_get_field constructs a function which returns the proper value"""
        import odict
        import awstats_cache_merge

        od = odict.OrderedDict([('pages', 4), ('hits', 15), ('bandwidth', 386873)])

        f = awstats_cache_merge.make_get_field('bandwidth')

        self.assertEqual(f(('dz', od)), 386873)


awstats_reader/__init__.py

"""
AwstatsReader
A library for querying statistics from AWStats (non-XML) Cache files
"""

import datetime
import glob
import os

# Requires odict from http://www.voidspace.org.uk/python/odict.html
import odict
od = odict.OrderedDict
d = dict

class AwstatsDateTime(datetime.datetime):
    """
    A subclass of datetime.datetime to handle AWStats' usage of
    year '0' in some cases.
    """
    def strftime(self, format):
        # A special flag value used by Awstats Reader
        # It's an (relatively safe) assumption that Awstats Reader won't be
        # processing data from year 1 AD
        # This is used because there are cases where an AWSTats date
        # (or datetime) field may be simply '0'
        if self.year == 1:
            return '0'
        else:
            return datetime.datetime.strftime(self, format)

class AwstatsDate(datetime.date):
    pass

def awstats_datetime(date_string):
    """
    Parses an AWStats Date/Time or Date string and returns a datetime.datetime
    or datetime.date object, respectively.
    """
    date_len = len(date_string)
    if date_len == 14:
        return AwstatsDateTime(int(date_string[0:4]), int(date_string[4:6]), int(date_string[6:8]),
                                 int(date_string[8:10]), int(date_string[10:12]),int(date_string[12:14]))

    elif date_len == 8:
        return AwstatsDate(int(date_string[0:4]), int(date_string[4:6]), int(date_string[6:8]))
    elif  date_len == 1 and date_string == '0':
        # Used if the year is zero. Dates are always compared, never added
        return AwstatsDateTime(1,1,1)
    else:
        raise RuntimeError("Invalid date/time string: '%s'" % date_string)

class AttrDict(odict.OrderedDict):
    """
    Allows dicts to be accessed via dot notation as well as subscripts
    """
    def __getattr__(self, name):
        return self[name]

class AwstatsReader(object):
    """
    The top-level object that takes the directory and domain and finds all
    the cache files, and returns the year objects given a subscript of a year
    """
    years = property(lambda self:self.__year_list)

    def __init__(self, directory, domain):
        self.__directory = directory
        self.__domain = domain
        self.__years = {}
        self.__year_list = []
        self.__curr_year_index = -1

        if not os.path.exists(directory):
            raise OSError((2, 'No such directory: %s' % directory, directory))

        for fname in glob.glob(os.path.join(directory, 'awstats??????.' + domain + '.txt')):
            stat_name = os.path.basename(fname)
            year = int(stat_name[9:13])
            month = int(stat_name[7:9])
            if year not in self.__years:
                self.__years[year] = AwstatsYear(self.__domain, year)

            self.__years[year]._set_month(month, AwstatsMonth(year, month, fname))

        self.__year_list = sorted(self.__years.keys())

    def __getitem__(self, year):
        try:
            return self.__years[year]
        except KeyError:
            raise KeyError("Domain '%s' does not have any records for year %s" % (self.__domain, year))

    def __contains__(self, year):
        return year in self.__year_list

    def __iter__(self):
        return (self.__years[y] for y in self.__year_list)

    def __str__(self):
        return "<AwstatsReader: " + ', '.join([str(y) for y in self.__year_list]) + ">"

class AwstatsYear(object):
    """
    The AWStats object containing the months for a given year
    """
    def __init__(self, domain, year):
        self.__domain = domain
        self.__year = year
        self.__months = {}
        self.__month_list = []

    def _set_month(self, month, mobject):
        self.__months[month] = mobject
        self.__month_list.append(month)

    def __iter__(self):
        return (self.__months[m] for m in sorted(self.__months.keys()))

    def __contains__(self, month):
        return month in self.__month_list

    def __getitem__(self,  month):
        try:
            return self.__months[month]
        except KeyError:
            raise KeyError("Domain '%s' does not have any records for year %s, month %s"
                           % (self.__domain, self.__year, month))

    def __str__(self):
        self.__month_list = sorted(self.__months.keys())
        return "<AwstatsYear " + str(self.__year) + ": " + ', '.join([str(m) for m in self.__month_list]) + ">"

    months = property(lambda self:self.__month_list)
    year = property(lambda self:self.__year)

class AwstatsMonth(object):
    """
    The AWStats object containing the sections for a given month
    """
    def __init__(self, year, month, fname):
        self.__year = year
        self.__month = month
        self.__version = None
        self.__fname = fname
        self.__pos_map = {}
        self.__section_list = []
        self.__section_cache = {}
        self.__fobject = None

    def __init_file(self):
        self.__fobject = open(self.__fname)

        version = self.__fobject.readline().split()[3:6]
        self.__version = (version[0], version[2].replace(')',''))

        for line in self.__fobject:
            if line.startswith('POS_'):
                # The Map lines (and others for that matter) have trailing spaces
                # Truly odd
                k,v = line.split(' ', 1)
                k = k[4:].lower()
                self.__pos_map[k] = int(v)
                self.__section_list.append(k)
            if line.startswith('END_MAP'):
                break

    def __get_raw_section(self, name):
        section_data = odict.OrderedDict()
        k = v = ''
        end_flag = 'END_' + name.upper()
        self.__fobject.seek(self.__pos_map[name])
        lines = int(self.__fobject.readline().split(' ')[1])
        for x in xrange(lines):
            line_data = self.__fobject.readline().strip()
            # Seems to be an off-by-one error in some sections
            if line_data == end_flag:
                break
            k,v = line_data.split(' ', 1)
            section_data[k] = v.split(' ')
        return section_data

    def __get_section(self, name):
        if not self.__fobject:
            self.__init_file()
        try:
            if name not in self.__section_cache:
                self.__section_cache[name] = AwstatsSection(self.__version, name, self.__get_raw_section(name))
            return self.__section_cache[name]
        except KeyError:
            raise KeyError("Section '%s' does not exist" % name)

    def __iter__(self):
        """
        Iterates through the list of sections in the month
        """
        if not self.__fobject:
            self.__init_file()
        return (s for s in self.__section_list)

    def __len__(self):
        """
        Returns the number of sections in the month
        """
        if not self.__fobject:
            self.__init_file()
        return len(self.__section_list)

    def __str__(self):
        return "<AwstatsMonth " + str(self.__year) + "-" + str(self.__month).rjust(2, '0') +">"

    __getitem__ = __get_section
    __getattr__ = __get_section

    def keys(self):
        if not self.__fobject:
            self.__init_file()
        return self.__section_list

    version = property(lambda self:self.__version)
    year = property(lambda self:self.__year)
    month = property(lambda self:self.__month)


class AwstatsSection(object):
    """
    Object containing the data from a section in a month's file
    """
    def __init__(self, version, section_name, raw_data):
        self.__name = section_name
        self.__format = _section_format['__default__'][section_name]
        self.__data = raw_data

    def __str__(self):
        return "<AwstatsSection %s, %s>" % (self.__name, self.__data)

    def __get_data(self, row_name):
        data = self.__data[row_name]
        if row_name in self.__format:
            format = self.__format[row_name]
        else:
            format = self.__format['__default__']

        if isinstance(format, tuple):
            items = odict.OrderedDict()
            for index, f in enumerate(format):
                if len(f) == 3 and f[2] == 'opt' and len(data) <= index:
                    break
                items[f[0]] = f[1](data[index])
            return AttrDict(items)
        else:
            return format(data[0])

    __getitem__ = __get_data
    __getattr__ = __get_data

    def keys(self):
        return list(self.__iter__())

    def get(self, name, default=None):
        try:
            return self.__get_data(name)
        except KeyError:
            return default

    def __iter__(self):
        return (k for k in self.__data.keys())

    def __len__(self):
        return len(self.__data)

    def items(self):
        return ((k,self.__data[k]) for k in self.__data.keys())

    def get_sort_info(self):
        sort_num = None
        sort_by = None
        sort_reversed = None

        if '__meta__' in self.__format:
            sort_num = self.__format['__meta__'].get('sort', None)
            sort_by = self.__format['__meta__'].get('sortby', None)
            sort_reversed = self.__format['__meta__'].get('reversed', True)

        return (sort_num, sort_by, sort_reversed)

    def __merge_sum(v1, v2):
        return v1 + v2

    def __merge_min(v1, v2):
        return min(v1, v2)

    def __merge_max(v1, v2):
        return max(v1, v2)

    def __merge_latest(v1, v2):
        """
        Right now, I'm assuming that the second set of files specified will
        be the later files. I have an idea for making this better, but haven't
        had time to really think it through and code it up.
        """
        return v2

    merge_funcs = {'sum':__merge_sum, 'min':__merge_min,
                   'max':__merge_max, 'latest':__merge_latest}

    def merge(self, other, row_name, field_name):
        """
        'other' is the AwstatsSection object with which we are merging
        """
        if row_name in _section_merge_rules['__default__'][self.__name]:
            merge_rule = _section_merge_rules['__default__'][self.__name][row_name][field_name]
        else:
            merge_rule = _section_merge_rules['__default__'][self.__name]['__default__'][field_name]

        if merge_rule.startswith('repl'):
            c,v = merge_rule.split(':',1)
            return str(v)
        else:
            try:
                return AwstatsSection.merge_funcs[merge_rule](self.get(row_name)[field_name],
                                                              other.get(row_name)[field_name])
            except KeyError:
                raise RuntimeError("Unhandled merge rule for section '%s', row '%s', field '%s': '%s'"
                                   % (self.__name, row_name, field_name, merge_rule))


_section_format = {}

_section_format['__default__'] = {
    'general':{
        'LastLine':(('date',awstats_datetime),('line',int),('offset',long),('signature',long)),
        'FirstTime':(('first_time', awstats_datetime),),
        'LastTime':(('last_time',awstats_datetime),),
        'LastUpdate':(('date',awstats_datetime),('parsed',int),('old',int),('new',int),('corrupted',int),('dropped',int)),
        '__default__':(('value', int),),
        },
    'time':{'__default__':(('pages',int),('hits',int),('bandwidth',int),('not_viewed_pages',int),
                       ('not_viewed_hits',int),('not_viewed_bandwidth',int)),
            '__meta__':{'sort':24, 'sortby':'key_int', 'sort_reversed':False}},
    'visitor':{'__default__':(('pages',int),('hits',int),('bandwidth',int),('last_visit',awstats_datetime,'opt'),
                              ('last_visit_start',awstats_datetime,'opt'),('last_visit_page',str,'opt')),
               '__meta__':{'sort':25, 'sortby':'pages'}},
    'day':{'__default__':(('pages',int),('hits',int),('bandwidth',int),('visits',int)),
           '__meta__':{'sort':31, 'sortby':'key', 'sort_reversed':False}},
    'domain':{'__default__':(('pages',int),('hits',int),('bandwidth',int)),
              '__meta__':{'sort':25, 'sortby':'pages'}},
    'login':{'__default__':(('pages',int),('hits',int),('bandwidth',int),('last_visit',awstats_datetime)),
             '__meta__':{'sort':10, 'sortby':'pages'}},
    'robot':{'__default__':(('hits',int),('bandwidth',int),('last_visit',awstats_datetime),('hits_on_robots',int)),
             '__meta__':{'sort':25, 'sortby':'hits'}},
    'worms':{'__default__':(('hits',int),('bandwidth',int),('last_visit',awstats_datetime)),
             '__meta__':{'sort':5, 'sortby':'hits'}},
    'emailsender':{'__default__':(('hits',int),('bandwidth',int),('last_visit',awstats_datetime)),
                   '__meta__':{'sort':20, 'sortby':'hits'}},
    'emailreceiver':{'__default__':(('hits',int),('bandwidth',int),('last_visit',awstats_datetime)),
            '__meta__':{'sort':20, 'sortby':'hits'}},
    'session':{'__default__':(('value', int),)},
    'sider':{'__default__':(('pages',int),('bandwidth',int),('entry',int),('exit',int)),
            '__meta__':{'sort':25, 'sortby':'pages'}},
    'filetypes':{'__default__':(('hits',int),('bandwidth',int),('bandwidth_without_compression',int),
                                ('bandwidth_after_compression',int))},
    'os':{'__default__':(('value', int),)},
    'browser':{'__default__':(('value', int),)},
    'screensize':{'__default__':(('value', int),)},
    'unknownreferer':{'__default__':(('value', awstats_datetime),)},
    'unknownrefererbrowser':{'__default__':(('value', awstats_datetime),)},
    'origin':{'__default__':(('pages',int),('hits',int)),
              '__meta__':{'sort':5, 'sortby':'key'}},
    'sereferrals':{'__default__':(('pages',int),('hits',int))},
    'pagerefs':{'__default__':(('pages',int),('hits',int)),
                '__meta__':{'sort':25, 'sortby':'pages'}},
    'searchwords':{'__default__':(('value', int),),
                   '__meta__':{'sort':10, 'sortby':'value'}},
    'keywords':{'__default__':(('value', int),),
                '__meta__':{'sort':25, 'sortby':'value'}},
    'misc':{'__default__':(('pages',int),('hits',int),('bandwidth',int))},
    'errors':{'__default__':(('hits',int),('bandwidth',int)),
              '__meta__':{'sort':5, 'sortby':'key_int'}},
    'cluster':{'__default__':(('pages',int),('hits',int),('bandwidth',int))},
    'sider_404':{'__default__':(('hits',int),('last_url_referer',str))},
    'plugin_geoip_city_maxmind':{'__default__':(('pages',int),('hits',int),('bandwidth',int),('last_access',awstats_datetime))},
}

"""
Merges data from two months.  The Rules, as far as I can document from
an AWSTats cache file.
"""
_section_merge_rules = {}
_section_merge_rules['__default__'] = {
    'general':{
        'LastLine':d((('date','max'),('line','max'),('offset','max'),('signature','repl:'))),
        'FirstTime':{'first_time':'min'},
        'LastTime':{'last_time':'max'},
        'LastUpdate':d((('date','max'),('parsed','sum'),('old','sum'),('new','sum'),('corrupted','sum'),('dropped','sum'))),
        '__default__':{'value':'sum'},
        },
    'time':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum'),('not_viewed_pages','sum'),
                       ('not_viewed_hits','sum'),('not_viewed_bandwidth','sum')))},
    'visitor':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum'),('last_visit','max'),
                              ('last_visit_start','max'),('last_visit_page','max')))},
    'day':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum'),('visits','sum')))},
    'domain':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum')))},
    'login':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum'),('last_visit','max')))},
    'robot':{'__default__':d((('hits','sum'),('bandwidth','sum'),('last_visit','max'),('hits_on_robots','sum')))},
    'worms':{'__default__':d((('hits','sum'),('bandwidth','sum'),('last_visit','max')))},
    'emailsender':{'__default__':d((('hits','sum'),('bandwidth','sum'),('last_visit','max')))},
    'emailreceiver':{'__default__':d((('hits','sum'),('bandwidth','sum'),('last_visit','max')))},
    'session':{'__default__':{'value':'sum'}},
    'sider':{'__default__':d((('pages','sum'),('bandwidth','sum'),('entry','sum'),('exit','sum')))},
    'filetypes':{'__default__':d((('hits','sum'),('bandwidth','sum'),('bandwidth_without_compression','sum'),
                                ('bandwidth_after_compression','sum')))},
    'os':{'__default__':{'value':'sum'}},
    'browser':{'__default__':{'value':'sum'}},
    'screensize':{'__default__':{'value':'sum'}},
    'unknownreferer':{'__default__':{'value':'max'}},
    'unknownrefererbrowser':{'__default__':{'value':'max'}},
    'origin':{'__default__':d((('pages','sum'),('hits','sum')))},
    'sereferrals':{'__default__':d((('pages','sum'),('hits','sum')))},
    'pagerefs':{'__default__':d((('pages','sum'),('hits','sum')))},
    'searchwords':{'__default__':{'value':'sum'}},
    'keywords':{'__default__':{'value':'sum'}},
    'misc':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum')))},
    'errors':{'__default__':d((('hits','sum'),('bandwidth','sum')))},
    'cluster':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum')))},
    'sider_404':{'__default__':d((('hits','sum'),('last_url_referer','latest')))},
    'plugin_geoip_city_maxmind':{'__default__':d((('pages','sum'),('hits','sum'),('bandwidth','sum'),('last_access','max')))},
}
